### PR TITLE
saver_child should not inherit xss-lock's sleep lock

### PR DESCRIPTION
**Pull request by piec**
Hi,

Thanks for making xsecurelock it's great :+1: 

I noticed that I was hitting systemd's inhibitor timeout when using it with xss-lock's sleep lock transfer.
Actually the lock is inherited by the xsecurelock'saver process when forking.
This patch closes the lock in the child process after forking so that the lock is released correctly.

This patch is a quick fix, tell me if you want me to change things

Cheers,
Pierre

---
More info:
* What you see in the logs if the lock is not released:
```
systemd-logind[14711]: Delay lock is active (UID 1000, PID 5243/n/a) but inhibitor timeout is reached.
```
* The systemd-logind timeout value is set with [`InhibitDelayMaxSec`](https://www.freedesktop.org/software/systemd/man/logind.conf.html#InhibitDelayMaxSec=) (default value 5s)
* By default systemd ignores the inhibitors when suspending because of the lid. [`LidSwitchIgnoreInhibited` default value](https://www.freedesktop.org/software/systemd/man/logind.conf.html#PowerKeyIgnoreInhibited=) is yes

### CODE REVIEW

File: saver_child.c [hunk 1]
```diff
@@ -97,6 +99,14 @@ void WatchSaverChild(Display* dpy, Window w, int index, const char* executable,
...
     } else if (pid == 0) {
+      int fd = GetIntSetting("XSS_SLEEP_LOCK_FD", -1);
+      unsetenv("XSS_SLEEP_LOCK_FD");
+      if (fd > 0) {
+        close(fd);
+        LogErrno("close");
+        // even if closing the lock fails we will reach systemd's timeout
+        // see InhibitDelayMaxSec
+      }
       // Child process.
```
Comment (schachmat) [inline saver_child.c]: You should move that comment to the top of the "else if" branch.

**Comment by piec** [comment]
Thanks, I just updated the patch

**Comment by divVerent** [comment]
Basically good, but it looks like this might not always prevent a similar issue with the auth child (assuming one suspends, then accidentally invokes the auth dialog too early). Not sure if this can currently be triggered, but it doesn't seem safe.

I wonder if it'd be better to simply set the  FD_CLOEXEC flag on the FD.

**Comment by piec** [comment]
True, I forgot about FD_CLOEXEC when I did that, that will definitely be better

**Comment by divVerent** [comment]
Great - can you do that? If not, I may find time for that in a few days.

**Comment by piec** [comment]
Yes I'll take a look

**Comment by piec** [comment]
oops, I reverted my fork master so github automatically closed the MR

**Comment by piec** [comment]
oops, I reverted my fork master so github automatically closed the MR

**Comment by divVerent** [comment]
Thanks, looks good and harmless!

File: main.c [hunk 2]
```diff
@@ -398,6 +399,16 @@ void NotifyOfLock(int x11_fd) {
 int main(int argc, char **argv) {
   setlocale(LC_CTYPE, "");
 
+  int fd = GetIntSetting("XSS_SLEEP_LOCK_FD", -1);
+  if (fd >= 0) {
+    int flags=  fcntl(fd, F_GETFD);
```
Comment (schachmat) [inline main.c]: s/=  / = /

File: main.c [hunk 3]
```diff
@@ -398,6 +399,16 @@ void NotifyOfLock(int x11_fd) {
...
+    if (flags == 1)
+      /* Handle error */;
```
Comment (schachmat) [inline main.c]: I think printing a warning in this and the next case of fcntl failure should be good. It's not critical, so we don't have to abort.

**Comment by piec** [comment]
I just updated the patch with some comments + proper logging on erros

**Pull request merged**